### ISSUE TITLE
[Feature Request] Mockup environment to run graph::query_trace

### ISSUE THREAD

**Issue by mbezuljTT** [tenstorrent/tt-metal#14000]
**Is your feature request related to a problem? Please describe.**

In the new mlir compiler, we want to use graph capturing/tracing in order to capture per op L1 allocation and to validate op arguments and hyper parameters.

With that in mind, for model compilation without TT hardware attached, it's important to support being able to run graph capturing/tracing in the mockup environment. Something like this

```

       auto call = [&] {
            const auto output_tensor = ttnn::add(input_tensor_a, input_tensor_b);
            return output_tensor;
        };

        auto json_trace = graph::query_trace(call);

```

**Describe the solution you'd like**

I've already got tt_mockup_device in the umd https://github.com/tenstorrent/tt-umd/pull/66

Real minimum to get this going is in this draft PR https://github.com/tenstorrent/tt-metal/pull/13998

It runs with `TT_METAL_MOCKUP_EN=1 ARCH_NAME=wormhole_b0 TT_METAL_SLOW_DISPATCH_MODE=1 ./build_Debug/test/ttnn/unit_tests_ttnn --gtest_filter=AddOpGraphTests/AddOpGraphTestFixture.AddGraphTrace/0` on a pure compute machine with metal docker. 

Formally I've had to select slow dispatch, because fast dispatch path casts tt_device to tt_silicon_device multiple times. While this could be rerouted to tt_metal_device - it would be dead code anyway - because mockup device doesnt do much.

Ideally, we would want all dispatch code off in this path. Because all dispatcher related compilation is not really being used and ultimately slows down things.

Let's discuss here or meet over zoom next week on what's proper way you would like to have this portion resolved. We're happy to push changes forward if metal team is not staffed at the moment.  

**Describe alternatives you've considered**
As bare minimum we can stay with formally calling TT_METAL_SLOW_DISPATCH_MODE=1  and few special exist under Device::initialize_and_launch_firmware() and Device::build_firmware() for TT_METAL_MOCKUP_EN=1. With these we have been able to get to ~1ms per query

**Comment by staylorTT** [comment]
@prajaramanTT This issue should be P1 based off our review from this week. Without this feature we have no workaround.

**Comment by skhatriTT** [comment]
@mbezuljTT Do we have an ETA for this?

**Comment by mbezuljTT** [comment]
I'm not the right person to ask that question. 

Feature request is for metal runtime @pgkeller , priority to be agreed with compiler folks @staylorTT

**Comment by pgkeller** [comment]
No eta, this work is not scheduled

**Comment by zzigler-tt** [comment]
@pgkeller should we get this FR added to the roadmap to provide a little more transparency on when the team can get around to this request?

**Comment by odjuricicTT** [comment]
@zzigler-tt That would we great. This enables Forge to compile without having a physical device and also increases performance forge compiled models can achieve.

**Comment by skhatriTT** [comment]
@pgkeller would it be possible to get a rough timeline for when this might be scheduled for?

**Comment by pgkeller** [comment]
Progress on this is contingent on runtime's hiring plan as present staffing can't address it.  We currently have 4 offers out to expand our team, the 2nd set of two will be the staffing to address this (assuming accepts).  ETA to begin work is still a couple months away

**Comment by skhatriTT** [comment]
@pgkeller checking in for an update

**Comment by abhullar-tt** [comment]
> [@pgkeller](https://github.com/pgkeller) checking in for an update

Haven't started this but this planned to be picked up end of this year/early next year

**Comment by odjuricicTT** [comment]
@abhullar-tt Is it possible for this to be picked up earlier?
It is creating a significant amount of pain on our side:
* Having this would prevent a whole type of hang bugs that we keep encountering with graph capture and spending a lot of time debugging (this is preventing us from brining up customer models)
* It would decrease forge compile time
* It would allow and enable users to run forge compilation on a machine without a physical TT device present

**Comment by abhullar-tt** [comment]
> [@abhullar-tt](https://github.com/abhullar-tt) Is it possible for this to be picked up earlier? It is creating a significant amount of pain on our side:
> 
> * Having this would prevent a whole type of hang bugs that we keep encountering with graph capture and spending a lot of time debugging (this is preventing us from brining up customer models)
> * It would decrease forge compile time
> * It would allow and enable users to run forge compilation on a machine without a physical TT device present

We now have a functional simulator. Would these flows want to run on that or will they act as if there is no TT backend at all?

**Comment by mbezuljTT** [comment]
(still getting notifications as I made original feature request though I'm not directly involved)


out of curiosity @abhullar-tt are you referring to versim-like simulator or something else? if it's something else (and faster) I'm very interested into it

**Comment by abhullar-tt** [comment]
> (still getting notifications as I made original feature request though I'm not directly involved)
> 
> out of curiosity [@abhullar-tt](https://github.com/abhullar-tt) are you referring to versim-like simulator or something else? if it's something else (and faster) I'm very interested into it

tt-sim repo: https://github.com/tenstorrent/ttsim
not versim like and would be slower than silicon but can simulate the entire chip. There is also WIP support for multi-chip. For the purposes of this issue I was wondering if it makes sense to make graph capture target tt-sim (@ayerofieiev-tt for your thoughts as well).

**Comment by odjuricicTT** [comment]
> We now have a functional simulator. Would these flows want to run on that or will they act as if there is no TT backend at all?

@abhullar-tt 
The main requirement of these flows is that we are able to call ttnn ops in graph capture NO_DISPATCH mode (without kernel compilation and dispatch) without having a device present on the machine. So having a simulator seem like overkill, since there is no need for simulating hardware.

**Comment by ayerofieiev-tt** [comment]
I highly recommend trying to run with ttsim as it can be used immediately. 
Keep in mind that ttsim is already a prioritiy for the company and it has strong xfn support. 
ttsim is not needed to run the workload, but it basically acts as a virtual device.

**Comment by ayerofieiev-tt** [comment]
It takes 5 minuted to try it. You build it. You set env variable. And now you can run metal on a system w/o real device. 
It has gaps, some of them are being closed fast. Others won't impact you given you run in no dispatch mode.

**Comment by abhullar-tt** [comment]
Seconding that, even if tt-sim provides richer features than what is needed here it essentially provides the interface needed here. One suggestion to make this more seamless is to have graph trace automatically use tt-sim / have Runtime expose a cleaner way to use tt-sim compared to env var approach

**Comment by mbezuljTT** [comment]
wow I had no idea tt-sim exists!

**Comment by pgkeller** [comment]
confused.  tt-sim is great, but this feature request is still valid even w/ tt-sim.  it is still valuable to run through a full workload w/o actually executing something, it'll run dramatically faster (and on smaller systems) than simulation.  right?

**Comment by abhullar-tt** [comment]
> confused. tt-sim is great, but this feature request is still valid even w/ tt-sim. it is still valuable to run through a full workload w/o actually executing something, it'll run dramatically faster (and on smaller systems) than simulation. right?

If I understand correctly, Graph trace already skips execution but atm it needs to be backed by actual HW. If Graph trace is only enabled on top of a Device that is backed by tt-sim then there isn't really a need to stub out any of the current codepaths

**Comment by pgkeller** [comment]
> > confused. tt-sim is great, but this feature request is still valid even w/ tt-sim. it is still valuable to run through a full workload w/o actually executing something, it'll run dramatically faster (and on smaller systems) than simulation. right?
> 
> If I understand correctly, Graph trace already skips execution but atm it needs to be backed by actual HW. If Graph trace is only enabled on top of a Device that is backed by tt-sim then there isn't really a need to stub out any of the current codepaths

ok, I get it.  that sounds more to me like a way to make forward progress rather than a final solution, right?  ie, we still want to enable running on a pure virtual device.  This will come in handy for offline builds as well

**Comment by odjuricicTT** [comment]
Agreed with @pgkeller, we could use tt-sim to make progress on our end, but it would not be a great fit for a final solution.

**Comment by skhatriTT** [comment]
@pgkeller do we have a timeline for when this issue will be picked up? What are the next steps here?

**Comment by abhullar-tt** [comment]
@skhatriTT this is being worked on by @msudumbrekar-TT, he will allow users to specify that their backend is a Mock ClusterType and existing apis will work without needing to be changed.

**Comment by odjuricicTT** [comment]
@abhullar-tt Awesome! Thanks you!
Is there an initial ETA on when this could land?

**Comment by msudumbrekar-TT** [comment]
@odjuricicTT The initial ETA im assuming is 2 weeks! Will try and prioritize it to finish it earlier if possible!

**Comment by odjuricicTT** [comment]
@abhullar-tt @msudumbrekar-TT Hey folks, any progress made on this issue?

**Comment by abhullar-tt** [comment]
> [@abhullar-tt](https://github.com/abhullar-tt) [@msudumbrekar-TT](https://github.com/msudumbrekar-TT) Hey folks, any progress made on this issue?

@msudumbrekar-TT made progress to enable slow dispatch through mock device but got taken off this full time to investigate another higher priority ask before he finished enabling fast dispatch. Will keep you posted on progress here

**Comment by skhatriTT** [comment]
Hey @abhullar-tt, could you please provide an updated ETA?

**Comment by abhullar-tt** [comment]
> Hey [@abhullar-tt](https://github.com/abhullar-tt), could you please provide an updated ETA?

Sorry, there was a higher priority task that interrupted @msudumbrekar-TT progress on this and is requiring active debug. We are hoping to get back to this once initial debug triage is underway

**Comment by odjuricicTT** [comment]
Hey @abhullar-tt any updates on if the work on this can be resumed?

**Comment by abhullar-tt** [comment]
> Hey [@abhullar-tt](https://github.com/abhullar-tt) any updates on if the work on this can be resumed?

yep @msudumbrekar-TT is unblocked to pick this work up again ( I can let him give better eta on FastDispatch support)

**Comment by odjuricicTT** [comment]
@abhullar-tt  @msudumbrekar-TT  Great!!! Is there an updated ETA?

**Comment by msudumbrekar-TT** [comment]
Hey @odjuricicTT id estimate it to be around 2 weeks!

**Comment by skhatriTT** [comment]
@msudumbrekar-TT can you please provide an update? Thank you!

**Comment by skhatriTT** [comment]
Quick update from a discussion with @msudumbrekar-TT - the mock environment is working for most chips in slow dispatch mode. Work is currently ongoing to enable support for fast dispatch as well.

Thank you @msudumbrekar-TT!

**Comment by msudumbrekar-TT** [comment]
Hey folks! Sorry for the delay, the feature has been tested locally and will probably get merged by early next week. This is the draft PR just for your reference: https://github.com/tenstorrent/tt-metal/pull/34194

**Comment by skhatriTT** [comment]
Hey @msudumbrekar-TT can this issue be closed?

**Comment by rpavlovicTT** [comment]
Great progress @msudumbrekar-TT ! I tried running `./build/test/ttnn/unit_tests_ttnn --gtest_filter="*QueryOpConstraints*"` with your changes and things look good!

--- 

I have a question regarding setting `mock` mode. Currently, it is done through env var. 
`tt-mlir` would prefer if it can be done programatically. 

Note that we use this API for opening a device today `::tt::tt_metal::distributed::MeshDevice::create`. I understand your flow with `rtoptions` triggers in lower parts of the stack, but it'd be great if we could **avoid** controlling this feature through choosing appropriate yaml file. Maybe we can expose API from tt-metal that lets us create mock device yaml, for example:
```
namespace tt::tt_metal {
    // Explicit arch specification
    void configure_mock_mode(tt::ARCH arch, uint32_t num_chips = 1);
    
    // Auto-detect from hardware (uses same arch/topology as physical device)
    void configure_mock_mode_from_hw();
}
```

So we would have easy way to choose mock device configuration, and also we could just rely on underlying HW and keep current code working as is, with just adding one additional call to `configure_mock_mode_from_hw()` before we create `MeshDevice`.

**Comment by odjuricicTT** [comment]
> Hey [@msudumbrekar-TT](https://github.com/msudumbrekar-TT) can this issue be closed?

@skhatriTT No, the issue is dependent on this PR https://github.com/tenstorrent/tt-metal/pull/34194 which is still active.

**Comment by pgkeller** [comment]
@rpavlovicTT Note that we have (stalled) work on updating the Metal Context to be an explicit object that will customize the device environment.  From 10K feet, mock device would be a part of this.  This is unfortunately not staffed at the moment, but we will figure it out in the coming week post holiday.  Is the env var an ok stop-gap solution?  Or do we need to address this w/ greater urgency?

**Comment by rpavlovicTT** [comment]
@pgkeller 
Honestly I would avoid env var approach since that means we would have to find appropriate cluster descriptor from the place of usage of `MockDevice`, which can be either compiler backend `tt-mlir` or compiler frontend like `tt-xla`. And cluster descriptors are currently hidden from layers above tt-metal. It just complicates things. I think we can wait an extra week if that means this will be resolved.

**Comment by pgkeller** [comment]
ok.  the diff will be more than a week.  we'll try to get this pinned down soon now that everyone is back from break

**Comment by skhatriTT** [comment]
Hey @pgkeller thanks for pushing it. Could you please provide an updated eta it would be really helpful?

**Comment by abhullar-tt** [comment]
> Hey [@pgkeller](https://github.com/pgkeller) thanks for pushing it. Could you please provide an updated eta it would be really helpful?

Should be merged today! Yesterday was the goal but the merge queue was backed up/broken and infra team suggested to wait until today

**Comment by msudumbrekar-TT** [comment]
Hey folks! Mock device PR #34194 has been merged! 

This implementation addresses all requirements from this issue:
Programmatic API (no env vars needed):
- configure_mock_mode(arch, num_chips) - explicit configuration
- configure_mock_mode_from_hw() - auto-detect from hardware
- disable_mock_mode() - turn off mock mode

Usage:
Option 1 - Programmatic API:
tt::tt_metal::experimental::configure_mock_mode(tt::ARCH::WORMHOLE_B0, 8);
auto mesh_device = tt::tt_metal::distributed::MeshDevice::create(...);

Quick Testing:
./build_Release/test/tt_metal/unit_tests_device --gtest_filter="MockDeviceAPIFixture.*"

Option 2 - Environment variable (for testing/CI):
export TT_METAL_MOCK_CLUSTER_DESC_PATH=/path/to/cluster_desc.yaml

Testing script for quick validation: ./tests/scripts/run_mock_device_tests.sh tt_metal/third_party/umd/tests/cluster_descriptor_examples/blackhole_P100.yaml (just change yaml cluster)

This unblocks tt-mlir graph capture without hardware! Thanks!

**Comment by rpavlovicTT** [comment]
I am integrating the mock device API (from commit 468b7e790cfea3452d2d740cb38fdc7fc5be731b) into tt-mlir for hardware-free compilation and testing. I have a question about the behavior of disable_mock_mode().  
                                                                                                                                                                                                                      
  Observed Behavior:                                                                                                                                                                                                  
                                                                                                                                                                                                                      
  1. We call configure_mock_mode(WORMHOLE_B0, 1) before MeshDevice::create() - works correctly                                                                                                                        
  2. After closing the device, we call disable_mock_mode()                                                                                                                                                            
  3. is_mock_mode_registered() returns false as expected                                                                                                                                                              
  4. However, when we create a new device (without calling configure_mock_mode()), it's still a mock device                                                                                                           
                                                                                                                                                                                                                      
  The logs show:                                                                                                                                                                                                      
```
  --- Phase B: Real Hardware ---                                                                                                                                                                                      
  is_mock_mode_registered() = false                                                                                                                                                                                   
  Creating real device (no mock mode)...                                                                                                                                                                              
  ... Skipping firmware initialization for mock devices   <-- still mock!                                                                                                                                             
```                                                                                                                                                                                                            


###  Root Cause Analysis:                                                                                                                                                                                               
                                                                                                                                                                                                                      
  Looking at the tt-metal code, I found that disable_mock_mode() clears the programmatic config flag, but the mock cluster descriptor loaded into rtoptions_ persists. Once MetalContext is initialized with mock  mode, subsequent device creations continue using the mock cluster descriptor.                                                                                                                                       
                                                                                                                                                                                                                      
  Question:                                                                                                                                                                                                           
                                                                                                                                                                                                                      
  Is this behavior intentional? We have use cases where we'd like to:                                                                                                                                                 
  1. Run optimizer passes with mock device (no HW needed for constraint queries)                                                                                                                                      
  2. Then switch to real hardware for actual execution in the same process                                                                                                                                            
                                                                                                                                                                                                                      
  Is it possible to fully reset mock mode and return to real hardware within the same process? If so, what's the recommended approach?                                                                                
                                                                                                                                                                                                                      
  Reproduction:                                                                                                                                                                                                       
                                                                                                                                                                                                                      
  Here's a minimal test case that demonstrates the behavior:                                                                                                                                                          
``` cpp                                                                                                                                                                                             
  #include "tt-metalium/experimental/mock_device.hpp"                                                                                                                                                                 
  #include "tt-metalium/host_api.hpp"                                                                                                                                                                                 
  #include "tt-metalium/mesh_device.hpp"                                                                                                                                                                              
  #include "ttnn/graph/graph_query_op_constraints.hpp"                                                                                                                                                                
  #include "ttnn/operations/normalization/softmax/softmax.hpp"                                                                                                                                                        
  #include "ttnn/tensor/tensor_spec.hpp"                                                                                                                                                                              
  #include <iostream>                                                                                                                                                                                                 
                                                                                                                                                                                                                      
  std::shared_ptr<::tt::tt_metal::distributed::MeshDevice> createDevice() {                                                                                                                                           
    size_t numDevices = ::tt::tt_metal::GetNumAvailableDevices();                                                                                                                                                     
    size_t numPCIeDevices = ::tt::tt_metal::GetNumPCIeDevices();                                                                                                                                                      
    ::tt::tt_metal::DispatchCoreType dispatchCoreType =                                                                                                                                                               
        numDevices == numPCIeDevices ? ::tt::tt_metal::DispatchCoreType::WORKER                                                                                                                                       
                                     : ::tt::tt_metal::DispatchCoreType::ETH;                                                                                                                                         
                                                                                                                                                                                                                      
    ::tt::tt_metal::distributed::MeshShape shape{1, 1};                                                                                                                                                               
    auto device = ::tt::tt_metal::distributed::MeshDevice::create(                                                                                                                                                    
        ::tt::tt_metal::distributed::MeshDeviceConfig{shape},                                                                                                                                                         
        32768, 6000000, 1, dispatchCoreType);                                                                                                                                                                         
    device->disable_and_clear_program_cache();                                                                                                                                                                        
    return device;                                                                                                                                                                                                    
  }                                                                                                                                                                                                                   
                                                                                                                                                                                                                      
  int main() {                                                                                                                                                                                                        
    // Phase A: Mock Device                                                                                                                                                                                           
    std::cout << "--- Phase A: Mock Device ---" << std::endl;                                                                                                                                                         
    ::tt::tt_metal::experimental::configure_mock_mode(::tt::ARCH::WORMHOLE_B0, 1);                                                                                                                                    
    auto mockDevice = createDevice();                                                                                                                                                                                 
    std::cout << "Mock device created" << std::endl;                                                                                                                                                                  
    mockDevice.reset();                                                                                                                                                                                               
    ::tt::tt_metal::experimental::disable_mock_mode();                                                                                                                                                                
                                                                                                                                                                                                                      
    // Phase B: Attempt Real Hardware                                                                                                                                                                                 
    std::cout << "\n--- Phase B: Real Hardware ---" << std::endl;                                                                                                                                                     
    std::cout << "is_mock_mode_registered() = "                                                                                                                                                                       
              << (::tt::tt_metal::experimental::is_mock_mode_registered() ? "true" : "false")                                                                                                                         
              << std::endl;                                                                                                                                                                                           
                                                                                                                                                                                                                      
    // This still creates a mock device despite not calling configure_mock_mode()                                                                                                                                     
    auto realDevice = createDevice();                                                                                                                                                                                 
    std::cout << "Device created (check logs - still mock?)" << std::endl;                                                                                                                                            
    realDevice.reset();                                                                                                                                                                                               
                                                                                                                                                                                                                      
    return 0;                                                                                                                                                                                                         
  }              
```

**Comment by abhullar-tt** [comment]
This seems like an issue with our MetalContext not being explicitly owned. @msudumbrekar-TT is addressing this with a workaround

**Comment by rpavlovicTT** [comment]
> This seems like an issue with our MetalContext not being explicitly owned. [@msudumbrekar-TT](https://github.com/msudumbrekar-TT) is addressing this with a workaround

Thanks for the quick response 🚀

**Comment by msudumbrekar-TT** [comment]
Hello team, i have addressed the above issue in this PR: https://github.com/tenstorrent/tt-metal/pull/36359. This workaround should solve the issue! Thanks!

**Comment by rpavlovicTT** [comment]
Thanks @msudumbrekar-TT , I will close this issue once we integrate this feature in tt-mlir and validate it. It looks good in tt-metal as of now, it will probably take few days to bubble up to tt-mlir.

**Comment by rpavlovicTT** [comment]
Feature is integrated, however we have another request regarding this feature to be able to use it fully. 

Problem: Mock devices cannot coexist with real devices

### Use Case

We need to create mock MeshDevices for graph compilation and planning while real devices are open (this is a limitation of our frontend which cannot be worked around):

1. Open a real MeshDevice
2. Create a mock MeshDevice alongside it (for graph compilation / buffer allocation planning)
3. Close the mock device
4. Continue using the real device

### Problem 1: `disable_mock_mode()` crashes with real devices open

`disable_mock_mode()` unconditionally calls `reinitialize_for_real_hardware()` whenever the device manager is initialized. It doesn't check whether MetalContext is actually running in mock mode.

If MetalContext was constructed *before* `configure_mock_mode()` was called, the singleton is already on real hardware — the mock config flag has no effect on it. Calling `reinitialize_for_real_hardware()` then hits a `TT_FATAL` because real devices are still active.

**Minimal fix:** Check `rtoptions().get_mock_enabled()` before calling `reinitialize_for_real_hardware()`. Skip the reinit if the context is already on real hardware. 

### Problem 2: No way to create mock devices alongside real devices

Even after fixing Problem 1, creating a mock MeshDevice while a real one is open is not supported:

- **MetalContext is a singleton** — either real mode or mock mode, not both. `configure_mock_mode()` sets a global flag that is only read during singleton construction.
- **`MeshDevice::create()` routes through MetalContext** — topology queries, device creation, and `SystemMesh` all operate on the single MetalContext instance.

There is no way to create a MeshDevice backed by mock infrastructure while MetalContext is serving real devices.

### What We Need

Support for creating mock MeshDevices independently of the real MetalContext, so mock and real devices can coexist. 

### Repro

```cpp
auto real_device = MeshDevice::create(MeshDeviceConfig(MeshShape{1, 1}));

experimental::configure_mock_mode(tt::ARCH::BLACKHOLE, 1);

// Tries to use real MetalContext — fails because chip 0 is already open
auto mock_device = MeshDevice::create(MeshDeviceConfig(MeshShape{1, 1}));  // CRASH

// Even without the above, this crashes:
experimental::disable_mock_mode();  // TT_FATAL: "Cannot switch to real hardware while 1 device(s) are active"
```

---- 

@abhullar-tt given the singleton constraint in MetalContext, what's the recommended path for supporting mock devices alongside real ones? Is this something the runtime team could support?
